### ISSUE TITLE
SIGBUS crash at "Restarting after configuration change."

### ISSUE THREAD

**Issue by Elias481** [Icinga/icinga2#7132]
## Describe the bug
It sometimes happens that icinga2 processes die/get killed with SIGBUS when they are doing "Restarting after configuration change".

Apr 19 16:27:38 hostname kernel: traps: icinga2[20625] trap stack segment ip:2b8778e7e61c sp:7fff02466f20 error:0 in libc-2.17.so[2b8778dfe000+1c2000]
Apr 19 16:27:38 hostname systemd: icinga2.service: main process exited, code=killed, status=7/BUS

## To Reproduce
?

## Expected behavior
No crashes at "Restarting after configuration change"

## Your Environment
* Version used (`icinga2 --version`): observer at least with 2.10.2, 2.10.3
* Operating System and version: Centos 7.6
* Enabled features (`icinga2 feature list`): api checker mainlog notification
* Config validation (`icinga2 daemon -C`): some sadly "unavoidable" `Apply rule does not match anywhere` warning from global-zone's apply rules...

## Additional context
There are not too many services monitored on the crashing instances and there is no evidence that this is related to special systems and happens quite seldom until now (thought icinga2 migration not completed and config not as often changed as it is expected to change somewhen).

Sadly I do not have any coredumps and there is no crash/error information collected/logged by icinga2 for these errors in our env.
I think last time there was no line regarding stack in system logs messages but I cannot say for sure.

**Comment by dnsmichi** [comment]
Not sure in which direction we should go here. Some pointers: https://stackoverflow.com/questions/2089167/debugging-sigbus-on-x86-linux

**Comment by Elias481** [comment]
I also found these pointers and some similiar, but I think without a core dump or at least some details gathered from such it's difficult to proceed. Possibly someone gets such a core and finds this issue when searching for the reason and can provide it?

I have to speak with my collegue whether he would accept to rollout and configure a mechanism to get core dumps from icinga2 on "all" our systems (as default behaviour is to not generate them because of suid-behaviour and a cwd without write permissiond for user I think one needs to install abrtd or configure the systemd-coredump utility on centos7 etc. - and in addition force abrtd to not deny processing because something it does not like regading packet signing...).
I did this only on "my" few systems where I did not see this one until now (at least not since abrtd is prepared).

Alternativley the whole reload procedure could be reviewed, but without a clue at which exact position the errorconditions happens this is at least much efford and possibly without success.

**Comment by dnsmichi** [comment]
When we have a core dump, we can analyse further. Since we've also analysed and fixed several crashes with your help already, I'm not sure whether these can be side effects and will only exist in 2.10 but fixed in future 2.11. I'll leave this open up until either you, we or someone else runs into this with more details.

**Comment by HOSTED-POWER** [comment]
We also experience random crashes on reload of the monitoring master. This is going on for quite a while now, we hoped this would have been fixed already :)

It only happens once every x days on 1 server at a time mostly.

We just experienced it again and could fine a crashlog:

  Application version: r2.10.4-1

System information:
  Platform: CentOS Linux
  Platform version: 7 (Core)
  Kernel: Linux
  Kernel version: 3.10.0-957.10.1.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid
Stacktrace:

        (0) libc.so.6: gsignal (+0x37) [0x7f5fdc9e9207]
        (1) libc.so.6: abort (+0x148) [0x7f5fdc9ea8f8]
        (2) libc.so.6: <unknown function> (+0x78d27) [0x7f5fdca2bd27]
        (3) libc.so.6: <unknown function> (+0x7f5d4) [0x7f5fdca325d4]
        (4) libc.so.6: <unknown function> (+0x82d98) [0x7f5fdca35d98]
        (5) libc.so.6: __libc_malloc (+0x4c) [0x7f5fdca381ac]
        (6) libstdc++.so.6: operator new(unsigned long) (+0x1d) [0x7f5fdd512ecd]
        (7) libstdc++.so.6: std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&) (+0x59) [0x7f5fdd571a19]
        (8) icinga2: char* std::string::_S_construct<char const*>(char const*, char const*, std::allocator<char> const&, std::forward_iterator_tag) (+0x21) [0x760891]
        (9) libstdc++.so.6: std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&) (+0x38) [0x7f5fdd5736d8]
        (10) icinga2: icinga::String::String(char const*) (+0x1e) [0xa887ee]
        (11) icinga2: icinga::JsonRpcConnection::MessageHandler(icinga::String const&) (+0x288) [0x8adb58]
        (12) icinga2: icinga::JsonRpcConnection::MessageHandlerWrapper(icinga::String const&) (+0x5b) [0x904f5b]
        (13) icinga2: icinga::WorkQueue::RunTaskFunction(std::function<void ()> const&) (+0x29) [0x757629]
        (14) icinga2: icinga::WorkQueue::WorkerThreadProc() (+0x5e7) [0x75e2f7]
        (15) libboost_thread-mt.so.1.53.0: <unknown function> (+0xd27a) [0x7f5fdecad27a]
        (16) libpthread.so.0: <unknown function> (+0x7dd5) [0x7f5fdcd87dd5]
        (17) libc.so.6: clone (+0x6d) [0x7f5fdcab0ead]

***
* This would indicate a runtime problem or configuration error. If you believe this is a bug in Icinga 2
* please submit a bug report at https://github.com/Icinga/icinga2 and include this stack trace as well as any other
* information that might be useful in order to reproduce this problem.
***

Failed to launch GDB: No such file or directory


Because it happens so rarely and each time on another server, it's very hard to enable proper debugging.  I hope the above could give some pointers :)

**Comment by Elias481** [comment]
I think these errors are related. (Just one time the libc from within process can catch the error and the other time the kernel does it from outside of the process.)
But I think the sigabort-cases (where the libc is catching some [memory] issue instead of the kernel) where we get the crash-information from icinga2 are not the one's that point directly to the root-cause. I hope a core-dump from a SIGBUS would do. (Or possibly from a SIGSEGV like reported by piperator in issue #6684 where a running icinga::Application::StartReloadProcess() is visible in provided abrt-backtrace so maybe we could ask him whether he still has that one that was generated there?)

If there would be a complete coredump this could help as one could see what other threads are currently doing, but as long as a signal handler is causing the process (the other threads) to keep running until signal handler finished, which I assume is the behaviour, this wouldn't help too much as we couldn't see the exact positions of the threads at time of crash.

**Comment by dnsmichi** [comment]
That malloc stacktrace directly runs into a segfault, and I am still wondering why this happens with String at this point. One point we've learned from #6737 ... there is something going on with std::bind/std::function and memory. It could be related to specific compiler versions too.

```
    (12) icinga2: icinga::JsonRpcConnection::MessageHandlerWrapper(icinga::String const&) (+0x5b) [0x904f5b]
    (13) icinga2: icinga::WorkQueue::RunTaskFunction(std::function<void ()> const&) (+0x29) [0x757629]
```

**Comment by HOSTED-POWER** [comment]
> I think these errors are related. (Just one time the libc from within process can catch the error and the other time the kernel does it from outside of the process.)
> But I think the sigabort-cases (where the libc is catching some [memory] issue instead of the kernel) where we get the crash-information from icinga2 are not the one's that point directly to the root-cause. I hope a core-dump from a SIGBUS would do. (Or possibly from a SIGSEGV like reported by piperator in issue #6684 where a running icinga::Application::StartReloadProcess() is visible in provided abrt-backtrace so could also be related?)
> 

We see indeed SIGSEGV errors.

Isn't the stacktrace at least giving some pointers? :)

**Comment by dnsmichi** [comment]
malloc() fails to allocate memory for the new string being created at this point. I've seen this before with the string and dictionary class too, typically this happens when a new event from the outside triggers a handler signal. Though either memory already is corrupted at this point, or no more available - the syslog trap should tell you about the exact error, fasttop and variants.

The more I see all these problems, I'm considering problems with std::bind and copy/move by value into function objects on call. Still, this is out of the scope for this specific issue.

**Comment by Elias481** [comment]
As far as I understand this stacktrace of an malloc SIGABRT alone doesn't reveal the underlying problem. The malloc tries to allocate some memory which was marked as free but is still in use somehow. But we do not get a clue which kind of object was at the memory-location (so even less which part of code caused a delete operation on an object which is still in use). But it sounds like a typical refcounting problem (refcount is incremented and decremented simultanously causing destructor to run while other thread assumes the object is still available and use it).
I assume the cause for SIGBUS is most probably a failing pointer-dereference (just reading through the possible causes for SIGBUS on Linux x86(_64)) which also could be explained such a way (a destructor run already and the other thread that did the refcount increment at same time tries to dereference the invalidated pointer).
And finally possible SIGSEGVs reported during reload are also a typical outcome of accessing parts of an already deleted object.

But the backtrace of the SIGABRT most likely shows a backtrace of an unrealted threads and only the backtraces of SIGSEGV and SIGBUS would point to one of the threads involved in the "delete while in use" or "use when already deleted" situation and what exactly that thread is doing.
So if there would be a complete coredump of such an ABRT it could be analyzed for what the other threads are currently doing which might help to understand the issue. But on the other hand that coredump might be taken to late because of icinga2's builtin ABRT handler to see it really clear, I assume.

**Comment by HOSTED-POWER** [comment]
just another report

 cat report.1556864702.456165
  Application version: r2.10.4-1

System information:
  Platform: Debian GNU/Linux
  Platform version: 9 (stretch)
  Kernel: Linux
  Kernel version: 4.9.0-8-amd64
  Architecture: x86_64

Build information:
  Compiler: GNU 6.3.0
  Build host: 602f071dcede

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid
Stacktrace:

        (0) libc.so.6: gsignal (+0xcf) [0x7f10fc0d9fff]
        (1) libc.so.6: abort (+0x16a) [0x7f10fc0db42a]
        (2) libc.so.6: <unknown function> (+0x70c00) [0x7f10fc117c00]
        (3) libc.so.6: <unknown function> (+0x76fc6) [0x7f10fc11dfc6]
        (4) libc.so.6: <unknown function> (+0x7780e) [0x7f10fc11e80e]
        (5) icinga2: std::_Function_base::_Base_manager<std::_Bind<std::_Mem_fn<void (icinga::Timer::*)()> (boost::intrusive_ptr<icinga::Timer>)> >::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation) (+0x38) [0x55b2d8cfa558]
        (6) icinga2: icinga::ThreadPool::WorkerThread::ThreadProc(icinga::ThreadPool::Queue&) (+0x4ca) [0x55b2d8cb838a]
        (7) libboost_thread.so.1.62.0: <unknown function> (+0x12116) [0x7f10fea39116]
        (8) libpthread.so.0: <unknown function> (+0x74a4) [0x7f10fda5b4a4]
        (9) libc.so.6: clone (+0x3f) [0x7f10fc18fd0f]

***
* This would indicate a runtime problem or configuration error. If you believe this is a bug in Icinga 2
* please submit a bug report at https://github.com/Icinga/icinga2 and include this stack trace as well as any other
* information that might be useful in order to reproduce this problem.
***

Failed to launch GDB: No such file or directory

How would I change my system to make it crashdump? I read documentation before, but it didn't seem to work. It's a Debian in this case, some other systems are CentOS

**Comment by dnsmichi** [comment]
@HOSTED-POWER Any chance you'll test the snapshot packages? They contain fixes, especially for the last stacktrace showing the timer call.

**Comment by HOSTED-POWER** [comment]
Hmm I would, but we have so many servers and it's only once in a while it happens, so we'd need to update them all to the snapshot packages and then wait, but not so easy to do :)

**Comment by Elias481** [comment]
@HOSTED-POWER as there are different reasons why coredumps are not generated by default, it's probably easiest to setup some core-handler as this address most (or all?) of the reasons for not generating coredumps by default.

There is a systemd-coredump handler available for systemd systems.
On Centos7 the binary is installed by default with systemd package but the handler is not activated. (The reason why I do not have such coredumps at hand...)
For EL8 the Beta has this issue fixed (and hopefully the the production releases from Redhat/Centos 8 as well) and the handler is by default activated.
For Debian (at least v9) and Ubuntu (at least v18.04) there is a separate package systemd-coredump which installs the binary and activated the handler by default.
So to activate the handler a sysctl like the followings need to be established:
ex. RH8:
```
# cat /usr/lib/sysctl.d/50-coredump.conf
kernel.core_pattern=|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h %e
```
ex. Deb9:
```
# cat /usr/lib/sysctl.d/50-coredump.conf
kernel.core_pattern=|/lib/systemd/systemd-coredump %P %u %g %s %t 9223372036854775808 %e
```
The unclear part (and I doen't find it documented anywhere) is why on the EL8 an additional %h(ostname) is between the max-size and the executable name. So I do not know/test how it would work on EL7. I can try it next week when I'm back in the office.

If the %c(ore-size-limit) syntax is used, addtionally the limit for the process should not be set too low (as far as I can see the default for icinga2 is unlimited which is fine) because possibly the systemd-coredump will deny bigger dumps(?). And one could check settings in `/etc/systemd/coredump.conf` where the core-dumps could be disabled (or conffigured to be sent to journal only where they can gert lost quickly if journals fills and rolls over).


Another way I know of are the bug reporting tools of the distros.

For Centos/Redhat there is abrtd with the abrt-addon-ccpp. The abrt-ccpp service installs the necessary core-patterns.
With default settings it did not work properly at our systems. We had to reconfigure it a bit to make it work (thought it should have worked without that for my understanding as icinga2 is from a package with PGP key).
```
OpenGPGCheck=no
ProcessUnpackaged=yes
```
(`kernel.core_pattern = |/usr/libexec/abrt-hook-ccpp %s %c %p %u %g %t e` on EL6 `kernel.core_pattern = |/usr/libexec/abrt-hook-ccpp %s %c %p %u %g %t e %P %I %h` on EL7)

I see on my personal server which runs at ubuntu 18.04 there is per default a similiar tool called apport installed. This at least worked out of the box. For me a drawback is that the dump is first saved base64 encoded in the report file from where it needs to be extracted with for example apport-unpack which threw an error but worked that far). (`kernel.core_pattern = |/usr/share/apport/apport %p %s %c %d %P`)

Possibly there is a similiar tool for non-Ubuntu Debian.

We can see here that the %c(ore-size-soft-limit) is passed to the tools but default is probably unlimited with systemd. As far as I see the abrtd on centos6 does work even with ulimit default softlimit of 0 (and hard-limit unlimited).

So You could first check with `systctl kernel.core_pattern` whether and which error handler is active on the system and if You see one is active You could test wheter it's working with simple kill like `kill -SEGV $(< /var/run/icinga2/icinga2.pid)` or such.

If You want the kernel to directly generate default core files You would have to:
- ensure the CWD of the Icinga2-Process is writable by the icinga2 User (by default it runs with CWD=/ when started as service which isn't writable for icinga of course) or the kernel.core_pattern is an absolute pathname (according to man it must be an absolute path for kernels >=3.7) where the icinga user has write permission to.
- check core-size limits and max file-size limits for process
- kernel.core_pattern must resolve to a non-empty name (with is a combination of kernel.core_pattern if it's not a pipe and possibly kernel.core_uses_pid which appends the .PID if pid is not already in pattern).
- the fs.suid_dumpabl sysctl must probably be set to 2 or the for security reasons less recommended 1 instead of default 0 to allow suid programs to be dumped
If all that set it should then also generate the cores...


_Not sure whether the sigaborts wll generate cores because of the built-in handler.
The handler mostly does not really work.
Even if it finds the installed gdb and is able to start it up, the gdb can possibly not attach to the process because of `kernel.yama.ptrace_scope = 1` or such nowadays default settings and icinga2 is not setting up the gdb process to allow attaching disregarding this settings.
Also not sure wheter SELinux will allow the ptracing when icinga2 runs confined._

_But that's a separate issue that one could look after at some time._

_I think it would be best that the crash-handler then also would get a core dump from gdb and exits in a way that the system isn't generating another one to have it not even later after error condition as at runtime there could be debug-symbols missing or specific lookups for variables wich could be needed for further investigation.
As far as I can see there is no simple one-shot execution possible for the gdb execution with explicit ptrace-allowance but the gdb process would need to be launched without attaching, configured to be allowed to attach with `prctl(PR_SET_PTRACER, debugger_pid, 0, 0, 0);` and after that the attach and dump could take place. Or the icinga2 must run with the CAP_SYS_PTRACE enabled and propagate that CAP to the gdb._

**Comment by Elias481** [comment]
The issue becomes a bit more pressuring in our enviroment (migration to icinga2 is proceeding so more activity / config changes and more monitored entdpoints).

I observed the problem also on master server after a reload today.
I only saw/recognised this specific error condition on systemd systems and the service is in state "reloading" during this crash, so the restart by Cenots7 systemd that we configured because of the other bugs that are not fixed with 2.10 does not kick in here (thats some kind of sytemd bug in my opinion but anyway).

I identified some systems where this problems happens more frequently and will ask responsible collegue for permission to roll out core dump config to get progress here.

**Comment by dnsmichi** [comment]
See my comment in #7301 - maybe this is related somehow?

**Comment by dnsmichi** [comment]
#7320 will likely help here, since the entire reload logic needs rework.

**Comment by Elias481** [comment]
I see that could indeed help without knowing exact root cause.
Today I got a dump.
```           PID: 21110 (icinga2)
           UID: 550 (icinga)
           GID: 1231 (icinga)
        Signal: 7 (BUS)
     Timestamp: Thu 2019-07-18 09:10:17 CEST (2h 19min ago)
  Command Line: /usr/lib64/icinga2/sbin/icinga2 --no-stack-rlimit daemon --close-stdio -e /var/log/icinga2/error.log --reload-internal 4222
    Executable: /usr/lib64/icinga2/sbin/icinga2
 Control Group: /system.slice/icinga2.service
          Unit: icinga2.service
         Slice: system.slice
       Boot ID: 4763160566ec497aa3c847fec30ebb3c
    Machine ID: d155c5c856d14d38a39c68d04375197a
      Hostname: xxxxxxxx.wcs.priv.nmop.de
      Coredump: /var/lib/systemd/coredump/core.icinga2.550.4763160566ec497aa3c847fec30ebb3c.21110.1563433817000000.xz
       Message: Process 21110 (icinga2) of user 550 dumped core.

                Stack trace of thread 21110:
                #0  0x00002b55f215261c malloc_consolidate (libc.so.6)
                #1  0x00002b55f2154295 _int_malloc (libc.so.6)
                #2  0x00002b55f21571ac malloc (libc.so.6)
                #3  0x00000000005db856 yajl_buf_append (icinga2)
                #4  0x000000000082b6c7 yajl_gen_map_open (icinga2)
                #5  0x00000000008482c7 _ZL6EncodeP10yajl_gen_tRKN6icinga5ValueE.82478.21492 (icinga2)
                #6  0x000000000084857d _ZN6icinga10JsonEncodeERKNS_5ValueEb (icinga2)
                #7  0x000000000084bbb1 _ZN6icinga7Process3RunERKSt8functionIFvRKNS_13ProcessResultEEE (icinga2)
                #8  0x000000000084d0ce _ZN6icinga11Application18StartReloadProcessEv (icinga2)
                #9  0x000000000084d383 _ZN6icinga11Application12RunEventLoopEv (icinga2)
                #10 0x000000000084e544 _ZN6icinga17IcingaApplication4MainEv (icinga2)
                #11 0x0000000000a10b74 _ZN6icinga11Application3RunEv (icinga2)
                #12 0x00000000009682c7 _ZNK6icinga13DaemonCommand3RunERKN5boost15program_options13variables_mapERKSt6vectorISsSaISsEE (icinga2)
                #13 0x00000000009719a3 _ZL4Mainv.15176.16432 (icinga2)
                #14 0x00000000005ae837 main (icinga2)
                #15 0x00002b55f20f43d5 __libc_start_main (libc.so.6)
                #16 0x00000000005bbc8e _start (icinga2)

                Stack trace of thread 21280:
                #0  0x00002b55f1ec1965 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x00000000007834bb _ZN5boost18condition_variable4waitERNS_11unique_lockINS_5mutexEEE (icinga2)
                #2  0x0000000000780e76 _ZN6icinga9WorkQueue16WorkerThreadProcEv (icinga2)
                #3  0x00002b55effa927a thread_proxy (libboost_thread-mt.so.1.53.0)
                #4  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #5  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21317:
                #0  0x00002b55f21c520d poll (libc.so.6)
                #1  0x000000000084b552 _ZN6icinga7Process12IOThreadProcEi (icinga2)
                #2  0x00002b55f174c070 n/a (libstdc++.so.6)
                #3  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #4  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21164:
                #0  0x00002b55f1ec1965 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x000000000085453b _ZN6icinga10ThreadPool12WorkerThread10ThreadProcERNS0_5QueueE (icinga2)
                #2  0x00002b55effa927a thread_proxy (libboost_thread-mt.so.1.53.0)
                #3  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #4  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21163:
                #0  0x00002b55f1ec1965 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x000000000085453b _ZN6icinga10ThreadPool12WorkerThread10ThreadProcERNS0_5QueueE (icinga2)
                #2  0x00002b55effa927a thread_proxy (libboost_thread-mt.so.1.53.0)
                #3  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #4  0x00002b55f21cfead __clone (libc.so.6)
                #4  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21207:
                #0  0x00002b55f1ec497d accept (libpthread.so.0)
                #1  0x0000000000a12512 _ZN6icinga6Socket6AcceptEv (icinga2)
                #2  0x0000000000a5ba57 _ZN6icinga11ApiListener18ListenerThreadProcERKN5boost13intrusive_ptrINS_6SocketEEE (icinga2)
                #3  0x00000000006dca94 _ZNSt6thread5_ImplISt12_Bind_simpleIFSt5_BindIFSt7_Mem_fnIMN6icinga11ApiListenerEFvRKN5boost13intrusive_ptrINS4_6SocketEEEEEPS5_NS7_INS4_9TcpSocketEEEEEvEEE6_M_runEv.local.10670 (icinga2)
                #4  0x00002b55f174c070 n/a (libstdc++.so.6)
                #5  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #6  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21319:
                #0  0x00002b55f21c520d poll (libc.so.6)
                #1  0x000000000084b552 _ZN6icinga7Process12IOThreadProcEi (icinga2)
                #2  0x00002b55f174c070 n/a (libstdc++.so.6)
                #3  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #4  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21182:
                #0  0x00002b55f1ec1d12 pthread_cond_timedwait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x00000000006ec26d _ZN5boost18condition_variable10timed_waitERNS_11unique_lockINS_5mutexEEERKNS_10posix_time5ptimeE.local.16068 (icinga2)
                #2  0x0000000000a10093 _ZN6icinga5Timer15TimerThreadProcEv (icinga2)
                #3  0x00002b55f174c070 n/a (libstdc++.so.6)
                #4  0x00002b55f1ebddd5 start_thread (libpthread.so.0)
                #5  0x00002b55f21cfead __clone (libc.so.6)

                Stack trace of thread 21208:
                #0  0x00002b55f1ec1965 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x00000000007834bb _ZN5bo
```

**Comment by dnsmichi** [comment]
```
       Message: Process 21110 (icinga2) of user 550 dumped core.

                Stack trace of thread 21110:
                #0  0x00002b55f215261c malloc_consolidate (libc.so.6)
                #1  0x00002b55f2154295 _int_malloc (libc.so.6)
                #2  0x00002b55f21571ac malloc (libc.so.6)
                #3  0x00000000005db856 yajl_buf_append (icinga2)
                #4  0x000000000082b6c7 yajl_gen_map_open (icinga2)
                #5  0x00000000008482c7 _ZL6EncodeP10yajl_gen_tRKN6icinga5ValueE.82478.21492 (icinga2)
                #6  0x000000000084857d _ZN6icinga10JsonEncodeERKNS_5ValueEb (icinga2)
                #7  0x000000000084bbb1 _ZN6icinga7Process3RunERKSt8functionIFvRKNS_13ProcessResultEEE (icinga2)
                #8  0x000000000084d0ce _ZN6icinga11Application18StartReloadProcessEv (icinga2)
                #9  0x000000000084d383 _ZN6icinga11Application12RunEventLoopEv (icinga2)
                #10 0x000000000084e544 _ZN6icinga17IcingaApplication4MainEv (icinga2)
                #11 0x0000000000a10b74 _ZN6icinga11Application3RunEv (icinga2)
                #12 0x00000000009682c7 _ZNK6icinga13DaemonCommand3RunERKN5boost15program_options13variables_mapERKSt6vectorISsSaISsEE (icinga2)
                #13 0x00000000009719a3 _ZL4Mainv.15176.16432 (icinga2)
                #14 0x00000000005ae837 main (icinga2)
                #15 0x00002b55f20f43d5 __libc_start_main (libc.so.6)
                #16 0x00000000005bbc8e _start (icinga2)
```

points to #6684 which we will solve with 2.11 and replacing the JSON library. The 2.11 RC is planned for next week, would be good if you can test in then :)

**Comment by Elias481** [comment]
Yes I remember that issue. And I already stated that I don't see Yayl causing the issues but probably some other issue is causing the Yayl code to crash.
There is no clue that the Yayl lib is doing something wrong here or is problemtic.
The yayl lib does nothing more then mallocate an initial 2048 byte chunk for the json map (dictionary).
```#0  malloc_consolidate (av=av@entry=0x2b55f2498760 <main_arena>) at malloc.c:4147
4147              nextsize = chunksize(nextchunk);
(gdb) bt
#0  malloc_consolidate (av=av@entry=0x2b55f2498760 <main_arena>) at malloc.c:4147
#1  0x00002b55f2154295 in _int_malloc (av=av@entry=0x2b55f2498760 <main_arena>, bytes=bytes@entry=2048) at malloc.c:3444
#2  0x00002b55f21571ac in __GI___libc_malloc (bytes=2048) at malloc.c:2905
#3  0x00000000005db856 in yajl_buf_ensure_available (want=1, buf=0x2507660) at yajl_buf.c:42
#4  yajl_buf_append (buf=<optimized out>, data=<optimized out>, len=<optimized out>, len=<optimized out>, data=<optimized out>, buf=<optimized out>) at yajl_buf.c:73
#5  0x000000000082b6c7 in yajl_gen_map_open (g=g@entry=0x2a975a0) at yajl_gen.c:303
#6  0x00000000008482c7 in EncodeDictionary (dict=..., handle=0x2a975a0) at ../base/json.cpp:58
#7  Encode (handle=handle@entry=0x2a975a0, value=...) at ../base/json.cpp:110
#8  0x000000000084857d in icinga::JsonEncode (value=..., pretty_print=pretty_print@entry=false) at ../base/json.cpp:145
#9  0x000000000084bbb1 in ProcessSpawn (fds=<synthetic pointer>, adjustPriority=<optimized out>, extraEnvironment=..., arguments=...) at ../base/process.cpp:388
#10 icinga::Process::Run(std::function<void (icinga::ProcessResult const&)> const&) (this=this@entry=0x29399d0, callback=...) at ../base/process.cpp:989
#11 0x000000000084d0ce in icinga::Application::StartReloadProcess (this=this@entry=0x2b5614000cb0) at ../base/application.cpp:426
#12 0x000000000084d383 in icinga::Application::RunEventLoop (this=this@entry=0x2b5614000cb0) at ../base/application.cpp:355
#13 0x000000000084e544 in icinga::IcingaApplication::Main (this=0x2b5614000cb0) at ../icinga/icingaapplication.cpp:119
#14 0x0000000000a10b74 in icinga::Application::Run() (this=0x2b5614000cb0) at ../base/application.cpp:1033
#15 0x00000000009682c7 in icinga::DaemonCommand::Run (this=<optimized out>, vm=..., ap=...) at ../cli/daemoncommand.cpp:320
#16 0x00000000009719a3 in Main () at icinga.cpp:681
#17 0x00000000005ae837 in main (argc=8, argv=0x7fffd0044808) at icinga.cpp:975
```

Investigating the glibc malloc error points to a heap corruption (caused by something before this Yayl call).
```
(gdb) frame 0
#0  malloc_consolidate (av=av@entry=0x2b55f2498760 <main_arena>) at malloc.c:4147
4147              nextsize = chunksize(nextchunk);
(gdb) info args
av = 0x2b55f2498760 <main_arena>
(gdb) info locals
fb = <optimized out>
maxfb = 0x2b55f24987b0 <main_arena+80>
p = 0x25dc778
nextp = 0x2f6e6962732f3261
unsorted_bin = 0x2b55f24987b8 <main_arena+88>
first_unsorted = <optimized out>
nextchunk = 0x676e69636b8cfbaa
size = 7453010309032129586
nextsize = <optimized out>
prevsize = <optimized out>
nextinuse = <optimized out>
bck = <optimized out>
fwd = <optimized out>
(gdb) p *av
$1 = {mutex = 1, flags = 1, fastbinsY = {0x0, 0x0, 0x0, 0x0, 0x0, 0x25d5790, 0x0, 0x0, 0x0, 0x0}, top = 0x2e251a0, last_remainder = 0x26a9040, bins = {0x25fe450, 0x2c1b300, 0x26a9040,
    0x2a2e9e0, 0x2717870, 0x2c4aeb0, 0x26c9ca0, 0x2513f10, 0x2703030, 0x271b1e0, 0x2919cd0, 0x267b360, 0x25df550, 0x25f9eb0, 0x26f7290, 0x2609da0, 0x2b55f2498838 <main_arena+216>,
    0x2b55f2498838 <main_arena+216>, 0x2b55f2498848 <main_arena+232>, 0x2b55f2498848 <main_arena+232>, 0x26a6c30, 0x26ab210, 0x2440f40, 0x2448f60, 0x24df8a0, 0x2c91a90, 0x2513ff0,
    0x25b6760, 0x27f78a0, 0x27cf550, 0x2b00530, 0x2b09f90, 0x2a32200, 0x259c6c0, 0x2c91bd0, 0x2517dd0, 0x2a11ac0, 0x28e6e40, 0x2a0e640, 0x2a0e640, 0x20a7790, 0x20a7790,
    0x2b55f2498908 <main_arena+424>, 0x2b55f2498908 <main_arena+424>, 0x2a674a0, 0x2a674a0, 0x271a8a0, 0x25ffe00, 0x250a760, 0x22bc770, 0x273f3c0, 0x273f3c0,
    0x2b55f2498958 <main_arena+504>, 0x2b55f2498958 <main_arena+504>, 0x271be40, 0x25dbb40, 0x24e8470, 0x2a16910, 0x25da560, 0x25da560, 0x1ffc630, 0x25a8280, 0x294fff0, 0x2a89a80,
    0x1fa23d0, 0x1fa23d0, 0x2000710, 0x2000710, 0x2a339c0, 0x259de30, 0x292bb90, 0x292bb90, 0x2b55f24989f8 <main_arena+664>, 0x2b55f24989f8 <main_arena+664>,
    0x2b55f2498a08 <main_arena+680>, 0x2b55f2498a08 <main_arena+680>, 0x2a7cb70, 0x25f8500, 0x2a174b0, 0x2606840, 0x2dfe630, 0x25d79a0, 0x2b55f2498a48 <main_arena+744>,
    0x2b55f2498a48 <main_arena+744>, 0x2b55f2498a58 <main_arena+760>, 0x2b55f2498a58 <main_arena+760>, 0x2b55f2498a68 <main_arena+776>, 0x2b55f2498a68 <main_arena+776>,
    0x2b55f2498a78 <main_arena+792>, 0x2b55f2498a78 <main_arena+792>, 0x2b55f2498a88 <main_arena+808>, 0x2b55f2498a88 <main_arena+808>, 0x2b55f2498a98 <main_arena+824>,
    0x2b55f2498a98 <main_arena+824>, 0x2b55f2498aa8 <main_arena+840>, 0x2b55f2498aa8 <main_arena+840>, 0x2b55f2498ab8 <main_arena+856>, 0x2b55f2498ab8 <main_arena+856>, 0x267aa40,
    0x26fec40, 0x2b55f2498ad8 <main_arena+888>, 0x2b55f2498ad8 <main_arena+888>, 0x2b55f2498ae8 <main_arena+904>, 0x2b55f2498ae8 <main_arena+904>, 0x2b55f2498af8 <main_arena+920>,
    0x2b55f2498af8 <main_arena+920>, 0x2b55f2498b08 <main_arena+936>, 0x2b55f2498b08 <main_arena+936>, 0x269e940, 0x2dfdeb0, 0x2be7c20, 0x2be7c20, 0x2668540, 0x282f300,
    0x2b55f2498b48 <main_arena+1000>, 0x2b55f2498b48 <main_arena+1000>, 0x2a12f50, 0x2a12f50, 0x2511b40, 0x2a90030, 0x2b55f2498b78 <main_arena+1048>, 0x2b55f2498b78 <main_arena+1048>,
    0x2576590, 0x2576590, 0x2507ac0, 0x2a7d3d0, 0x2c90510, 0x25abbb0, 0x296dc90, 0x26c9ea0, 0x295d110, 0x268f500, 0x25d8b60, 0x26bb520, 0x2937390, 0x26ce050, 0x2a34d10, 0x25a94b0,
    0x2581f00, 0x29587d0, 0x29622a0, 0x295b050, 0x26b1200, 0x2593a60, 0x2925310, 0x293f070, 0x2b55f2498c48 <main_arena+1256>, 0x2b55f2498c48 <main_arena+1256>,
    0x2b55f2498c58 <main_arena+1272>, 0x2b55f2498c58 <main_arena+1272>, 0x2b55f2498c68 <main_arena+1288>, 0x2b55f2498c68 <main_arena+1288>, 0x2b55f2498c78 <main_arena+1304>,
    0x2b55f2498c78 <main_arena+1304>, 0x2b55f2498c88 <main_arena+1320>, 0x2b55f2498c88 <main_arena+1320>, 0x2b55f2498c98 <main_arena+1336>, 0x2b55f2498c98 <main_arena+1336>,
    0x2b55f2498ca8 <main_arena+1352>, 0x2b55f2498ca8 <main_arena+1352>, 0x2b55f2498cb8 <main_arena+1368>, 0x2b55f2498cb8 <main_arena+1368>, 0x2b55f2498cc8 <main_arena+1384>,
    0x2b55f2498cc8 <main_arena+1384>, 0x2b55f2498cd8 <main_arena+1400>, 0x2b55f2498cd8 <main_arena+1400>, 0x2b55f2498ce8 <main_arena+1416>, 0x2b55f2498ce8 <main_arena+1416>,
    0x2b55f2498cf8 <main_arena+1432>, 0x2b55f2498cf8 <main_arena+1432>, 0x2b55f2498d08 <main_arena+1448>, 0x2b55f2498d08 <main_arena+1448>, 0x2b55f2498d18 <main_arena+1464>,
    0x2b55f2498d18 <main_arena+1464>, 0x2b55f2498d28 <main_arena+1480>, 0x2b55f2498d28 <main_arena+1480>, 0x2b55f2498d38 <main_arena+1496>, 0x2b55f2498d38 <main_arena+1496>,
    0x2b55f2498d48 <main_arena+1512>, 0x2b55f2498d48 <main_arena+1512>, 0x2b55f2498d58 <main_arena+1528>, 0x2b55f2498d58 <main_arena+1528>, 0x2b55f2498d68 <main_arena+1544>,
    0x2b55f2498d68 <main_arena+1544>, 0x2b55f2498d78 <main_arena+1560>, 0x2b55f2498d78 <main_arena+1560>, 0x2b55f2498d88 <main_arena+1576>, 0x2b55f2498d88 <main_arena+1576>,
    0x2b55f2498d98 <main_arena+1592>, 0x2b55f2498d98 <main_arena+1592>, 0x2b55f2498da8 <main_arena+1608>, 0x2b55f2498da8 <main_arena+1608>, 0x2b55f2498db8 <main_arena+1624>,
    0x2b55f2498db8 <main_arena+1624>, 0x2b55f2498dc8 <main_arena+1640>, 0x2b55f2498dc8 <main_arena+1640>, 0x2b55f2498dd8 <main_arena+1656>, 0x2b55f2498dd8 <main_arena+1656>,
    0x2b55f2498de8 <main_arena+1672>, 0x2b55f2498de8 <main_arena+1672>...}, binmap = {4156553724, 3683059679, 66559, 1610614848}, next = 0x2b567c000020, next_free = 0x0,
  attached_threads = 1, system_mem = 16080896, max_system_mem = 16080896}
(gdb) p *p
$2 = {prev_size = 7091318039310988591, size = 7453010309032129590, fd = 0x2f6e6962732f3261, bk = 0x3261676e696369, fd_nextsize = 0x31, bk_nextsize = 0x0}
```
Digging a bit through malloc code one can find that the malloc_consolidate takes all the chunks from fastbin and consolidate them ("malloc_consolidate is a specialized version of free() that tears down chunks held in fastbins. ...", "Remove each chunk from fast bin and consolidate it, placing it then in unsorted bin. Among other reasons for doing this, placing in unsorted bin avoids needing to calculate actual bins until malloc is sure that chunks aren't immediately going to be reused anyway.")
That means it tries to reuse recently freed chunks for performance reasons.
One of these chunks in fastbin seems currupted. The "p" (the malloc_chunk* taken from fastbin) is, according to the data that would be recognised if such an (uncorrupted) struct is located at that address, a memory mapped file with a size of ‭931626288629016198‬ bytes (about 847310 TB). Then trying to read the footer (nextchunk) which each memmapped bin must have (for normal bins this overlpas with next bin from my understanding) yes it get's a SIGBUS trying to get that address...
```
(gdb) p *nextchunk
Cannot access memory at address 0x676e69636b8cfbaa
```
That's also explaining why it's a SIGBUS and not a SIGSEGV, it's outside the accesable memory and not different/wrong memory.

Such heap curruption must likely occurs if the preceding mallocated chunk was overused (so more data was written then the userdata could hold / more then requested while mallocationg that chunk) and thus overwriting the next chunks header information which can obviously lead to such an currupted bin.

So the question is: what is causing the heap corruption. It could be Yayl in some previous call, of course. But no point that far that this is really the case.

**Comment by Elias481** [comment]
Trying to find the previous chunks content one finds the following.
```
(gdb) p *(struct malloc_chunk *)(0x25dc5b0+112)
$71 = {prev_size = 1718513507, size = 65, fd = 0xb6b030 <vtable for icinga::EqualExpression+16>, bk = 0x25dc5d8, fd_nextsize = 0x1000000008, bk_nextsize = 0x2400000008}
(gdb) p/x (0x25dc5b0+112)
$72 = 0x25dc620
(gdb) p *(struct malloc_chunk *)(0x25dc620+64)
$73 = {prev_size = 0, size = 49, fd = 0xf, bk = 0xf, fd_nextsize = 0x5, bk_nextsize = 0x2d3261676e696369}
(gdb) p/x (0x25dc620+64)
$74 = 0x25dc660
(gdb) p *(struct malloc_chunk *)(0x25dc660+48)
$75 = {prev_size = 28259013886894440, size = 81, fd = 0x2600470, bk = 0x2600470, fd_nextsize = 0xffffffff, bk_nextsize = 0x6954796300000000}
(gdb) p/x (0x25dc660+48)
$76 = 0x25dc690
(gdb) p *(struct malloc_chunk *)(0x25dc690+80)
$77 = {prev_size = 80, size = 33, fd = 0x2502fb0, bk = 0x2503020, fd_nextsize = 0x0, bk_nextsize = 0x51}
(gdb) p/x (0x25dc690+80)
$78 = 0x25dc6e0
(gdb) p *(struct malloc_chunk *)(0x25dc6e0+32)
$79 = {prev_size = 0, size = 81, fd = 0xb628b0 <vtable for icinga::DictExpression+16>, bk = 0x25dc568, fd_nextsize = 0x2000000005, bk_nextsize = 0x100000009}
(gdb) p/x (0x25dc6e0+32)
$80 = 0x25dc700
(gdb) p *(struct malloc_chunk *)(0x25dc700+80)
$81 = {prev_size = 0, size = 65, fd = 0x1f, bk = 0x1f, fd_nextsize = 0x1, bk_nextsize = 0x62696c2f7273752f}
(gdb) p/x (0x25dc700+80)
$82 = 0x25dc750
(gdb) p *(struct malloc_chunk *)(0x25dc750+64)
$83 = {prev_size = 14180845697459049, size = 49, fd = 0x0, bk = 0x0, fd_nextsize = 0x1, bk_nextsize = 0x0}
(gdb) p/x (0x25dc750+64)
$90 = 0x25dc790
(gdb) p *(struct malloc_chunk *)(0x25dc790+48)
$93 = {prev_size = 0, size = 113, fd = 0x4c, bk = 0x4c, fd_nextsize = 0x4, bk_nextsize = 0x62696c2f7261762f}
(gdb) p/x (0x25dc790+48)
$94 = 0x25dc7c0
(gdb) p *(struct malloc_chunk *)(0x25dc7c0+112)
$95 = {prev_size = 1718513507, size = 177, fd = 0x25dc180, bk = 0x25ddf40, fd_nextsize = 0x1ffffffff, bk_nextsize = 0x6974617669746341}
(gdb) p/x (0x25dc7c0+112)
$96 = 0x25dc830
(gdb) p *(struct malloc_chunk *)(0x25dc830+176)
$97 = {prev_size = 176, size = 48,
  fd = 0xb49b50 <vtable for std::_Sp_counted_deleter<icinga::Expression*, std::default_delete<icinga::Expression>, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> [clone .local.19785]+16>, bk = 0x100000001, fd_nextsize = 0x25dc710, bk_nextsize = 0x0}
(gdb) p/x (0x25dc830+176)
$98 = 0x25dc8e0
```
That does not look like previous value has overflown but that the chunk in fastbin on 0x25dc778 just does not exists anymore. There is one on 0x25dc750 and on 0x25dc780, explaining the numbers/field contents when interpreting 0x25dc778 as `struct malloc_chunk *`..
So question is why is that fastbin there at that time.. I have no idea, but still do not see any clue that Yayl code put it there.

**Comment by Elias481** [comment]
Possibly it's just that C++ new and C malloc are not 100% compatibel / thread safe.
I think most allocations are done by new whichs memory allocation is implementation defined and can use same memory area as malloc. Both by them self are declared thread safe but nothing stated about mixed use of both has to be thread safe and such.

So probably removing yayl and possibly any other code that utlilizes malloc for heap memory allocation could mitigate these kind of errors.

**Comment by dnsmichi** [comment]
We'll see after 2.11 then again.

**Comment by Elias481** [comment]
We had no further crashes on some heavily affected machines since upgrading them to 2.11. So it looks like this problem is gone. Thanks.

**Comment by dnsmichi** [comment]
Thanks for the feedback 👍

**Comment by yoshi314** [comment]
i am seeing this right now on debian8 with icinga 2.11.4. 

I had to downgrade from 2.12, and this now keeps happening with increasing memory leakage unti icinga2 gets oom'd

**Comment by yoshi314** [comment]
```
*** Error in `/usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2': double free or corruption (out): 0x00007f25bc1982b0 ***
Caught SIGABRT.
Current time: 2020-08-12 20:59:21 +0200

*** Error in `/usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2': free(): corrupted unsorted chunks: 0x00007f9328132d80 ***
Caught SIGABRT.
Current time: 2020-08-13 13:52:42 +0200

*** Error in `/usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2': free(): corrupted unsorted chunks: 0x00007f93281b2860 ***
```
